README.md
Showdown's Youtube Extension
==========================

[![Build Status](https://travis-ci.org/showdownjs/youtube-extension.svg)](https://travis-ci.org/showdownjs/youtube-extension) [![npm version](https://badge.fury.io/js/showdown-youtube.svg)](http://badge.fury.io/js/showdown-youtube) [![npm version](https://badge.fury.io/bo/showdown-youtube.svg)](http://badge.fury.io/bo/showdown-youtube)

------

**An extension to embed [Youtube](http://youtube.com/) e [Vimeo](http://www.vimeo.com/) videos in showdown documents using markdown syntax**

## Introduction

This extension enables embedding youtube videos using markdown's image syntax. Instead of creating a new definition,
it borrows image syntax `![foo][bar.jpg]` to create an inframe pointing to the desired video.
Although it uses the same syntax as images, only valid youtube video links are actually parsed.
Basically it looks for urls that start with:
 - `http://www.youtube.com/watch?v=` or `youtube.com/watch?v=`
 - `http://www.youtube.com/embed/` or `youtube.com/embed/`
 - `http://youtu.be/` or `youtu.be/`
 - `http://vimeo.com/`

## Installation

### With [npm](http://npmjs.org)

    npm install showdown-youtube

### With [bower](http://bower.io/)

    bower install showdown-youtube

### Manual

You can also [download the latest release zip or tarball](https://github.com/showdownjs/youtube-extension/releases) and include it in your webpage, after showdown:

    <script src="showdown.min.js">
    <script src="showdown-youtube.min.js">

## Enabling the extension

After including the extension in your application, you just need to enable it in showdown.

    var converter = new showdown.Converter({extensions: ['youtube']});

When using in node, ensure to first require the extension so it can register itself with showdown before any converters try to use it.

```javascript
var showdown = require('showdown');
require('showdown-youtube');
var converter = new showdown.Converter({extensions: ['youtube']});
```

## Example

```javascript
var converter = new showdown.Converter({extensions: ['youtube']}),
    input = '![youtube video](http://www.youtube.com/watch?v=dQw4w9WgXcQ)';
    html = converter.makeHtml(input);
console.log(html);
```

This should output the equivalent to:

```html
<iframe src="//www.youtube.com/embed/dQw4w9WgXcQ?rel=0" frameborder="0" allowfullscreen></iframe>
```

## Video size

Since youtube extension uses the image syntax for videos, you can set the video dimensions the same way.

```md
![youtube video](http://www.youtube.com/watch?v=dQw4w9WgXcQ =800x600)
```

This should output the equivalent to:

```html
<iframe src="//www.youtube.com/embed/dQw4w9WgXcQ?rel=0" width="800px" height="600px" frameborder="0" allowfullscreen></iframe>
```

You can also use units:

```md
![foo](youtu.be/dQw4w9WgXcQ =100x80)     simple, assumes units are in px
![bar](youtu.be/dQw4w9WgXcQ =100x*)      sets the height to "auto"
![baz](youtu.be/dQw4w9WgXcQ =80%x5em)   width of 80% and height of 5em
```



## Options

Youtube extension reads the options from Showdown Core. So, to pass an option, use the methods explained in
[Showdown's documentation](https://github.com/showdownjs/showdown#setting-options).


### youtubeHeight
Sets the default height of the Youtube iframe.


### youtubeWidth
Sets the default width of the Youtube iframe.


### smoothLivePreview

(borrowed from Showdown core)

When using the extension in live editors (for instance, with angularjs), every time the text is changed, a new iframe is created.
This leads to a lot of requests done to youtube, which translates as an overall feel of sluggishness.
Turning **smoothLivePreview** on replaces the actual video objects (iframes) with an image placeholder. This way the live editing
feels more smooth.

This should only be enabled during live preview and not on the final document.


### youtubeUseSimpleImg
Only used if smoothLivePreview is on.

Uses a simple black image preview instead of an svg. This is quicker and ensures max compatibility with older browsers.

Default is false


## License
These files are distributed under BSD license. For more information,
please check the [LICENSE file](https://github.com/showdownjs/youtube-extension/blob/master/LICENSE) in the source code.


CHANGELOG.md
<a name="1.3.0"></a>
## [1.3.0](https://github.com/showdownjs/youtbe-extension/compare/1.2.1...v1.3.0)

#### Features
* Allow YouTube configuration to include `enablejsapi=1` into the final URL, which enables the JS api for YouTube videos.
  * [YouTube Docs](https://developers.google.com/youtube/player_parameters#enablejsapi)

<a name="1.2.1"></a>
## [1.2.1](https://github.com/showdownjs/youtbe-extension/compare/1.2.0...v1.2.1) (2016-11-29)

#### Features
* **defaultSizes:** allows a user to override the default height or width by passing in youtubeHeight or youtubeWidth
    into showdown settings ([26b220a](https://github.com/showdownjs/youtube-extension/commit/26b220a663a5bd480141304976f42eba215da75a))


<a name="1.2.0"></a>
## 1.2.0 (2016-11-23)

#### Fix
* **output:** force HTTPS protocol in output ([fa1d360](https://github.com/showdownjs/youtube-extension/commit/fa1d3608449a876f1e475e054c2c84fc070946b6))


<a name="1.1.0"></a>
## 1.1.0 (2016-06-21)

#### Features
* **vimeo:** Add vimeo support ([07d02e6](https://github.com/showdownjs/youtube-extension/commit/07d02e6db4f7ee5be6c88cffb9a0fb31d5ce9d8f))


<a name="1.0.0"></a>
## 1.0.0 (2015-07-21)

#### Release information

Release of youtube extension compatible with showdown ^1.0.0

#### Compatibility

**Compatible with showdown v.1.x.x**

#### Features

* **video size**: enables setting video size using image dimensions syntax.
* **smooth live preview**: enables smooth preview in live editors by replacing the iframe with an image. This option
    should be disabled when parsing the final document.


test/node.js
(function () {
  'use strict';

  require('chai').should();
  var showdown = require('showdown');
  require('../src/showdown-youtube.js');

  var fs = require('fs'),
      defaultConverter = new showdown.Converter({extensions: ['youtube']}),
      optionsConverter = new showdown.Converter({
        extensions: ['youtube'],
        youtubeHeight: 400,
        youtubeWidth: 400,
        youtubejsapi: true
      }),
      cases = fs.readdirSync('test/cases/')
        .filter(filter())
        .map(map('test/cases/')),
      issues = fs.readdirSync('test/issues/')
        .filter(filter())
        .map(map('test/issues/')),
      optionsCases = fs.readdirSync('test/optionsCases/')
        .filter(filter())
        .map(map('test/optionsCases/'));

  /////////////////////////////////////////////////////////////////////////////
  // Test cases
  //
  describe('Youtube Extension simple testcases', function () {
    for (var i = 0; i < cases.length; ++i) {
      it(cases[i].name, assertion(cases[i], defaultConverter));
    }
  });

  describe('Youtube Extension issues testcases', function () {
    for (var i = 0; i < issues.length; ++i) {
      it(issues[i].name, assertion(issues[i], defaultConverter));
    }
  });

  describe('Youtube Extension converter options testcases', function () {
    for (var i = 0; i < optionsCases.length; ++i) {
      it(optionsCases[i].name, assertion(optionsCases[i], optionsConverter));
    }
  });

  /////////////////////////////////////////////////////////////////////////////
  // Test cases
  //
  function filter() {
    return function (file) {
      var ext = file.slice(-3);
      return (ext === '.md');
    };
  }

  function map(dir) {
    return function (file) {
      var name = file.replace('.md', ''),
          htmlPath = dir + name + '.html',
          html = fs.readFileSync(htmlPath, 'utf8'),
          mdPath = dir + name + '.md',
          md = fs.readFileSync(mdPath, 'utf8');

      return {
        name:     name.replace(/_/g, ' '),
        input:    md,
        expected: html
      };
    };
  }

  //Normalize input/output
  function normalize(testCase) {

    // Normalize line returns
    testCase.expected = testCase.expected.replace(/\r/g, '');
    testCase.actual = testCase.actual.replace(/\r/g, '');

    // Ignore all leading/trailing whitespace
    testCase.expected = testCase.expected.split('\n').map(function (x) {
      return x.trim();
    }).join('\n');
    testCase.actual = testCase.actual.split('\n').map(function (x) {
      return x.trim();
    }).join('\n');

    // Remove extra lines
    testCase.expected = testCase.expected.trim();

    return testCase;

  }

  function assertion(testCase, converter) {
    return function () {
      testCase.actual = converter.makeHtml(testCase.input);
      testCase = normalize(testCase);

      // Compare
      testCase.actual.should.equal(testCase.expected);
    };
  }
})();


Gruntfile.js
module.exports = function (grunt) {
  'use strict';

  if (grunt.option('q') || grunt.option('quiet')) {
    require('quiet-grunt');
  }

  var config = {
    pkg: grunt.file.readJSON('package.json'),

    concat: {
      options: {
        banner: '/*! <%= pkg.name %> <%= grunt.template.today("dd-mm-yyyy") %> */\n',
        sourceMap: true
      },
      dist: {
        src: ['src/*.js'],
        dest: 'dist/<%= pkg.name %>.js'
      }
    },

    comments: {
      js: {
        options: {
          singleline: true,
          multiline: true
        },
        src: ['<%= concat.dist.dest %>']
      }
    },

    uglify: {
      options: {
        banner: '/*! <%= pkg.name %> <%= grunt.template.today("dd-mm-yyyy") %> */\n',
        sourceMap: true
      },
      dist: {
        files: {
          'dist/<%= pkg.name %>.min.js': ['<%= concat.dist.dest %>']
        }
      }
    },

    jshint: {
      options: {
        jshintrc: '.jshintrc',
        reporterOutput: ''
      },
      files: [
        'Gruntfile.js',
        'src/**/*.js',
        'test/**/*.js'
      ]
    },

    jscs: {
      options: {
        config: '.jscs.json'
      },
      files: [
        'Gruntfile.js',
        'src/**/*.js',
        'test/**/*.js'
      ]
    },

    conventionalChangelog: {
      options: {
        changelogOpts: {
          preset: 'angular'
        }
      },
      release: {
        src: 'CHANGELOG.md'
      }
    },

    conventionalGithubReleaser: {
      release: {
        options: {
          auth: {
            type: 'oauth',
            token: process.env.GH_TOEKN
          },
          changelogOpts: {
            preset: 'angular'
          }
        }
      }
    },

    //Server-side tests
    simplemocha: {
      test: {
        src: 'test/node.js',
        options: {
          globals:     ['should'],
          timeout:     3000,
          ignoreLeaks: false,
          reporter:    'spec'
        }
      }
    }
  };

  grunt.initConfig(config);

  require('load-grunt-tasks')(grunt);

  grunt.registerTask('lint', ['jshint', 'jscs']);
  grunt.registerTask('test', ['lint', 'simplemocha']);
  grunt.registerTask('build', ['test', 'concat', 'comments', 'uglify']);
  grunt.registerTask('prep-release', ['build', 'conventionalChangelog']);

  grunt.registerTask('default', ['build']);
};


src/youtube.js
/*  Youtube Extension
    ^http://www.youtube.com/watch?v=0mmx68VmTEo  ->
    <iframe  src=\"//www.youtube.com/embed/0mmx68VmTEo?rel=0\"\nframeborder=\"0\" allowfullscreen></iframe>
*/

(function() {
  var youtube = function(converter) {
    return [
      {
        type: 'lang',
        regex: '\\^\\^([\\S]+)',
        replace: function(match, url) {
          var m, video_id, youtube;
          youtube = /^.*((youtu.be\/)|(v\/)|(\/u\/\w\/)|(embed\/)|(watch\?))\??v?=?([^#\&\?]*).*/;
          if (youtube.test(url)) {
            m = url.match(youtube);
            if (m && m[7].length === 11) {
              video_id = m[7];
              return "<iframe  src=\"//www.youtube.com/embed/" + video_id + "?rel=0\"\nframeborder=\"0\" allowfullscreen></iframe>";
            } else {
              return match;
            }
          } else {
            return match;
          }
        }
      }
    ];
  };

  // Client-side export
  if (typeof window !== 'undefined' && window.Showdown && window.Showdown.extensions) { 
      window.Showdown.extensions.youtube = youtube;
  }
  // Server-side export
  if (typeof module !== 'undefined') module.exports = youtube;

}());


src/showdown-youtube.js
/**
 * Youtube Extension.
 * Uses image syntax to embed videos
 * Usage:
 * ![youtube video][http://youtu.be/dQw4w9WgXcQ]
 *
 * or
 *
 * ![youtube video][1]
 * [1]: http://youtu.be/dQw4w9WgXcQ
 */
(function (extension) {
  'use strict';

  if (typeof showdown !== 'undefined') {
    // global (browser or nodejs global)
    extension(showdown);
  } else if (typeof define === 'function' && define.amd) {
    // AMD
    define(['showdown'], extension);
  } else if (typeof exports === 'object') {
    // Node, CommonJS-like
    module.exports = extension(require('showdown'));
  } else {
    // showdown was not found so we throw
    throw Error('Could not find showdown library');
  }

}(function (showdown) {
  'use strict';

  var svg =
      '<div class="youtube-preview" style="width:%2; height:%3; background-color:#333; position:relative;">' +
      '<svg version="1.1" xmlns="http://www.w3.org/2000/svg" ' +
      '     width="100" height="70" viewBox="0 0 100 70"' +
      '     style="position: absolute; top: 50%; left: 50%; transform: translate(-50%, -50%);">' +
      '    <defs>' +
      '      <linearGradient id="grad1" x1="0%" x2="0%" y1="0%" y2="100%">' +
      '        <stop offset="0%" style="stop-color:rgb(229,45,49);stop-opacity:1" />' +
      '        <stop offset="100%" style="stop-color:rgb(191,23,29);stop-opacity:1" />' +
      '      </linearGradient>' +
      '    </defs>' +
      '    <rect width="100%" height="100%" rx="26" fill="url(#grad1)"/>' +
      '    <polygon points="35,20 70,35 35,50" fill="#fff"/>' +
      '    <polygon points="35,20 70,35 64,37 35,21" fill="#e8e0e0"/>' +
      '</svg>' +
      '<div style="text-align:center; padding-top:10px; color:#fff"><a href="%1">%1</a></div>' +
      '</div>',
    img = '<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs=" width="%2" height="%3">',
    iframe = '<iframe src="%1" width="%2" height="%3" frameborder="0" allowfullscreen></iframe>',
    imgRegex = /(?:<p>)?<img.*?src="(.+?)"(.*?)\/?>(?:<\/p>)?/gi,
    fullYoutubeRegex = /(?:(?:https?:)?(?:\/\/)?)(?:(?:www)?\.)?youtube\.(?:.+?)\/(?:(?:watch\?v=)|(?:embed\/))([a-zA-Z0-9_-]{11})/i,
    shortYoutubeRegex = /(?:(?:https?:)?(?:\/\/)?)?youtu\.be\/([a-zA-Z0-9_-]{11})/i,
    vimeoRegex = /(?:(?:https?:)?(?:\/\/)?)(?:(?:www)?\.)?vimeo.com\/(\d+)/;

  function parseDimensions(rest, options) {
    var width,
      height,
      d,
      defaultWidth,
      defaultHeight;

    defaultWidth = options.youtubeWidth ? options.youtubeWidth : 420;
    defaultHeight = options.youtubeHeight ? options.youtubeHeight : 315;

    if (rest) {
      width = (d = /width="(.+?)"/.exec(rest)) ? d[1] : defaultWidth;
      height = (d = /height="(.+?)"/.exec(rest)) ? d[1] : defaultHeight;
    }

    // add units so they can be used in css
    if (/^\d+$/gm.exec(width)) {
      width += 'px';
    }
    if (/^\d+$/gm.exec(height)) {
      height += 'px';
    }

    return {
      width: width,
      height: height
    };
  }

  /**
   * Replace with video iframes
   */
  showdown.extension('youtube', function () {
    return [
      {
        // It's a bit hackish but we let the core parsers replace the reference image for an image tag
        // then we replace the full img tag in the output with our iframe
        type: 'output',
        filter: function (text, converter, options) {
          var tag = iframe;
          if (options.smoothLivePreview) {
            tag = (options.youtubeUseSimpleImg) ? img : svg;
          }
          return text.replace(imgRegex, function (match, url, rest) {
            var d = parseDimensions(rest, options),
              m, fUrl = '';
            if ((m = shortYoutubeRegex.exec(url)) || (m = fullYoutubeRegex.exec(url))) {
              fUrl = 'https://www.youtube.com/embed/' + m[1] + '?rel=0';
              if (options.youtubejsapi) {
                fUrl += '&enablejsapi=1';
              }
            } else if ((m = vimeoRegex.exec(url))) {
              fUrl = 'https://player.vimeo.com/video/' + m[1];
            } else {
              return match;
            }
            return tag.replace(/%1/g, fUrl).replace('%2', d.width).replace('%3', d.height);
          });
        }
      }
    ];
  });
}));
